### PR TITLE
core: Add scalable dimensions to vector

### PR DESCRIPTION
**Pull request by math-fehr**
This is a feature that was missing in our current vector implementation.
I added support for it in the parser/printer and in the type.

### CODE REVIEW

**Review by superlopuh — APPROVED** [review]

File: xdsl/parser.py [hunk 1]
```diff
@@ -1279,28 +1285,38 @@ def try_parse_numerical_dims(self,
...
-        if self.tokenizer.next_token_of_pattern("[") is not None:
-            # We now need to parse the scalable dimensions
-            scaling_shape = list(self.try_parse_numerical_dims())
-            self.parse_characters(
-                "]", "Expected end of scalable vector dimensions here!")
-            self.parse_characters(
-                "x", "Expected end of scalable vector dimensions here!")
+        dims: list[int] = []
+        num_scalable_dims = 0
+        # First, we parse the static dimensions
```
Comment (georgebisbas) [inline xdsl/parser.py:1292]: can drop "we"

**Review by georgebisbas — COMMENTED** [review]
Can drop "we" from dosctrings

**Review by AntonLydike — APPROVED** [review]
This PR seems to add all the float stuff as well, just make sure to merge in the correct order ;)

**Comment by meshtag** [comment]
Sorry I couldn't respond to this PR during the week. 

In my opinion, we should also add tests which generate scalable vector IR via mlir. AFAIU, we are only checking things via the xDSL IR right now for scalable vector stuff, perhaps we should try generating mlir ir from it and add a filecheck test for that? 

I'm happy to do that if we agree it would be a good addition.

**Pull request merged**